src/cares_wrap.cc

```diff
@@ -1172,694 +580,622 @@ int ParseSoaReply(Environment* env,
+NodeAresTask* NodeAresTask::Create(ChannelWrap* channel, ares_socket_t sock) {
+  auto task = new NodeAresTask();
 
-    CallOnComplete(ret, ttls);
-  }
-};
+  task->channel = channel;
+  task->sock = sock;
 
-class QueryCaaWrap: public QueryWrap {
- public:
-  QueryCaaWrap(ChannelWrap* channel, Local<Object> req_wrap_obj)
-      : QueryWrap(channel, req_wrap_obj, "resolveCaa") {
+  if (uv_poll_init_socket(channel->env()->event_loop(),
+                          &task->poll_watcher, sock) < 0) {
+    /* This should never happen. */
+    delete task;
+    return nullptr;
   }
 
-  int Send(const char* name) override {
-    AresQuery(name, ns_c_in, T_CAA);
-    return 0;
-  }
+  return task;
+}
 
-  SET_NO_MEMORY_INFO()
-  SET_MEMORY_INFO_NAME(QueryAaaaWrap)
-  SET_SELF_SIZE(QueryAaaaWrap)
+void ChannelWrap::Setup() {
+  struct ares_options options;
+  memset(&options, 0, sizeof(options));
+  options.flags = ARES_FLAG_NOCHECKRESP;
+  options.sock_state_cb = ares_sockstate_cb;
+  options.sock_state_cb_data = this;
+  options.timeout = timeout_;
 
- protected:
-  void Parse(unsigned char* buf, int len) override {
-    HandleScope handle_scope(env()->isolate());
-    Context::Scope context_scope(env()->context());
+  int r;
+  if (!library_inited_) {
+    Mutex::ScopedLock lock(ares_library_mutex);
+    // Multiple calls to ares_library_init() increase a reference counter,
+    // so this is a no-op except for the first call to it.
+    r = ares_library_init(ARES_LIB_INIT_ALL);
+    if (r != ARES_SUCCESS)
+      return env()->ThrowError(ToErrorCodeString(r));
+  }
 
-    Local<Array> ret = Array::New(env()->isolate());
-    int status = ParseCaaReply(env(), buf, len, ret);
-    if (status != ARES_SUCCESS) {
-      ParseError(status);
-      return;
-    }
+  /* We do the call to ares_init_option for caller. */
+  const int optmask =
+      ARES_OPT_FLAGS | ARES_OPT_TIMEOUTMS | ARES_OPT_SOCK_STATE_CB;
+  r = ares_init_options(&channel_, &options, optmask);
 
-    this->CallOnComplete(ret);
+  if (r != ARES_SUCCESS) {
+    Mutex::ScopedLock lock(ares_library_mutex);
+    ares_library_cleanup();
+    return env()->ThrowError(ToErrorCodeString(r));
   }
-};
 
-class QueryCnameWrap: public QueryWrap {
- public:
-  QueryCnameWrap(ChannelWrap* channel, Local<Object> req_wrap_obj)
-      : QueryWrap(channel, req_wrap_obj, "resolveCname") {
-  }
+  library_inited_ = true;
+}
 
-  int Send(const char* name) override {
-    AresQuery(name, ns_c_in, ns_t_cname);
-    return 0;
+void ChannelWrap::StartTimer() {
+  if (timer_handle_ == nullptr) {
+    timer_handle_ = new uv_timer_t();
+    timer_handle_->data = static_cast<void*>(this);
+    uv_timer_init(env()->event_loop(), timer_handle_);
+  } else if (uv_is_active(reinterpret_cast<uv_handle_t*>(timer_handle_))) {
+    return;
   }
+  int timeout = timeout_;
+  if (timeout == 0) timeout = 1;
+  if (timeout < 0 || timeout > 1000) timeout = 1000;
+  uv_timer_start(timer_handle_, AresTimeout, timeout, timeout);
+}
```

Comment:
Can we move this timer logic in JS? Creating a lot of timers on libuv is somewhat expensive from other analysis.

Reply:
I would not move them unless we can actually prove that that’s worth it.

The reason that we switched the JS timers to not always create libuv timers was the cost of the JS/C++ boundary crossing, not inefficiency of libuv timers, iirc.

Reply:
That's true - however we are always creating these whenever we do a dns lookup, which we do all the time for every new socket created. Those are a lot of timers in addition to everything else.

There are probably better ways to cater for this, possibly deduplicating resolves or other tricks. Closing for now.

Reply:
Similar to the uv_poll_t, it's not likely going to be worth investing the time to refactor that if `getdns` is going to land.